### PR TITLE
Add SQL position to AST nodes

### PR DESCRIPTION
**Pull request by raghavsethi**
Pre-requisite to fixing #3203, which also requires making error messages friendlier.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/TupleAnalyzer.java [hunk 1]
```diff
@@ -460,7 +461,6 @@ protected TupleDescriptor visitJoin(Join node, AnalysisContext context)
         leftContext.setLateralTupleDescriptor(left);
         TupleDescriptor right = process(node.getRight(), leftContext);
 
-        // todo this check should be inside of TupleDescriptor.join and then remove the public getRelationAlias method, but the exception needs the node object
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/TupleAnalyzer.java]: Keep this TODO?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java [hunk 2]
```diff
@@ -246,9 +248,11 @@ public Void visitIndexJoin(IndexJoinNode node, Integer indent)
         {
             List<Expression> joinExpressions = new ArrayList<>();
             for (IndexJoinNode.EquiJoinClause clause : node.getCriteria()) {
-                joinExpressions.add(new ComparisonExpression(ComparisonExpression.Type.EQUAL,
-                        new QualifiedNameReference(clause.getProbe().toQualifiedName()),
-                        new QualifiedNameReference(clause.getIndex().toQualifiedName())));
+                joinExpressions.add(
+                        new ComparisonExpression(Optional.empty(),
+                        ComparisonExpression.Type.EQUAL,
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java]: Indentation is off for this line

**Comment by haozhun** [comment]
1. Should we require the parameter `Optional<Span>` on every constructor of subclasses of node?
2. Add tests that checks correct error location is in the message.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/SemanticException.java [hunk 3]
```diff
@@ -43,4 +47,14 @@ public SemanticErrorCode getCode()
     {
         return code;
     }
+
+    @Override
+    public String getMessage()
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/SemanticException.java]: Make a static method and call it in super.

File: presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java [hunk 4]
```diff
@@ -1317,4 +1353,46 @@ private static void check(boolean condition, String message, ParserRuleContext c
...
+                parserRuleContext.getStart().getLine(),
+                parserRuleContext.getStart().getStartIndex(),
+                parserRuleContext.getStop().getLine(),
+                parserRuleContext.getStop().getStopIndex());
+    }
+
+    public static Span fromRuleContext(List<? extends ParserRuleContext> parserRuleContexts)
+    {
+        requireNonNull(parserRuleContexts);
+        int startLineMin = MAX_VALUE;
```
Comment (haozhun) [inline presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java]: Don't static import this. It is pretty meaningless this way.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/ArithmeticUnaryExpression.java [hunk 5]
```diff
@@ -38,12 +51,12 @@ public ArithmeticUnaryExpression(Sign sign, Expression value)
 
     public static ArithmeticUnaryExpression positive(Expression value)
```
Comment (haozhun) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/ArithmeticUnaryExpression.java]: I believe this need to have SqlLocation. It's a factory method called directly from `AstBuilder`

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/ArithmeticUnaryExpression.java [hunk 6]
```diff
@@ -38,12 +51,12 @@ public ArithmeticUnaryExpression(Sign sign, Expression value)
...
     {
-        return new ArithmeticUnaryExpression(Sign.PLUS, value);
+        return new ArithmeticUnaryExpression(Optional.empty(), Sign.PLUS, value);
     }
 
     public static ArithmeticUnaryExpression negative(Expression value)
```
Comment (haozhun) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/ArithmeticUnaryExpression.java]: I believe this need to have SqlLocation. It's a factory method called directly from `AstBuilder`

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/SingleColumn.java [hunk 7]
```diff
@@ -24,8 +24,24 @@
     private final Optional<String> alias;
     private final Expression expression;
 
+    public SingleColumn(Expression expression)
```
Comment (haozhun) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/SingleColumn.java]: Why did you add this?

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/SingleColumn.java [hunk 8]
```diff
@@ -35,12 +51,17 @@ public SingleColumn(Expression expression, Optional<String> alias)
...
+        this(Optional.empty(), expression, Optional.of(alias));
     }
 
-    public SingleColumn(Expression expression)
+    private SingleColumn(Optional<Span> sqlLocation, Expression expression, String alias)
+    {
+        this(sqlLocation, expression, Optional.of(alias));
+    }
+
+    private SingleColumn(Optional<Span> sqlLocation, Expression expression)
```
Comment (haozhun) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/SingleColumn.java]: Remove

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/SingleColumn.java [hunk 9]
```diff
@@ -35,12 +51,17 @@ public SingleColumn(Expression expression, Optional<String> alias)
 
     public SingleColumn(Expression expression, String alias)
     {
-        this(expression, Optional.of(alias));
+        this(Optional.empty(), expression, Optional.of(alias));
     }
 
-    public SingleColumn(Expression expression)
+    private SingleColumn(Optional<Span> sqlLocation, Expression expression, String alias)
```
Comment (haozhun) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/SingleColumn.java]: Make `Optional`

**Comment by haozhun** [comment]
It looks good to me. @martint, please take a quick look.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/SemanticException.java [hunk 10]
```diff
@@ -23,10 +25,9 @@
     private final SemanticErrorCode code;
     private final Node node;
 
-    public SemanticException(SemanticErrorCode code, Node node, String format, Object... args)
+    public SemanticException(SemanticErrorCode code, Node node, String formatString, Object... args)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/SemanticException.java]: Why the name change?

**Comment by martint** [comment]
Wherever there are variables `sqlLocation`, please rename them to just `location`. Same for `getSqlLocation` -> `getLocation`.

**Comment by raghavsethi** [comment]
Overwritten comment by @martint: I think we need to track the start and end positions for the subtree of a node, in addition to the location of the corresponding operator. For instance, for:

SELECT ('a' || 'b') + (1 * 2) - 3 FROM t
The location of the ArithmeticBinaryExpression for the addition should be the location of the + sign, and the start and end should point to the starting ( and the 3, respectively. This will allow us to highlight the entire expression as having a problem, and point to the + as the operator having the type mismatch:

SELECT <<<('a' || 'b') + (1 * 2) - 3>>> FROM t
                       ^ 
Error at line 1:20. Type mismatch. Cannot add VARCHAR and BIGINT

**Comment by raghavsethi** [comment]
I was tracking the end locations earlier, but given that the downstream error handler only uses line and column numbers, @haozhun recommended getting rid of it. Should I put all of that in the same diff, or should I add this feature (by modifying SemanticException, ParseException and CLI) in another diff?

**Comment by martint** [comment]
It's fine for the exceptions to use the start position (for now), but the AST nodes should have the correct information.

**Comment by haozhun** [comment]
@martint You want completely unused fields in the code?

**Comment by martint** [comment]
> @martint You want completely unused fields in the code?

This is ok. It's a building block for later changes. My concern is that, otherwise, it may require a significant rewrite when we add it (hopefully soon).

**Comment by raghavsethi** [comment]
@martint: I think I got all the cases we discussed.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 11]
```diff
@@ -708,7 +710,19 @@ protected Type visitFunctionCall(FunctionCall node, AnalysisContext context)
...
-            Signature function = functionRegistry.resolveFunction(node.getName(), argumentTypes.build(), context.isApproximate());
+            Signature function;
+            try {
+                function = functionRegistry.resolveFunction(node.getName(), argumentTypes.build(), context.isApproximate());
+            }
+            catch (PrestoException e) {
+                if (e.getErrorCode().getCode() == StandardErrorCode.FUNCTION_NOT_FOUND.toErrorCode().getCode()) {
+                    throw new SemanticException(SemanticErrorCode.FUNCTION_NOT_FOUND, node, e.getMessage());
+                }
+                else {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: This `else` is unnecessary

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 12]
```diff
@@ -708,7 +710,19 @@ protected Type visitFunctionCall(FunctionCall node, AnalysisContext context)
                 argumentTypes.add(process(expression, context).getTypeSignature());
             }
 
-            Signature function = functionRegistry.resolveFunction(node.getName(), argumentTypes.build(), context.isApproximate());
+            Signature function;
+            try {
+                function = functionRegistry.resolveFunction(node.getName(), argumentTypes.build(), context.isApproximate());
+            }
+            catch (PrestoException e) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: At some point we should change the function registry to return an Optional and let the analyzer throw the appropriate exception.

**Comment by martint** [comment]
looks good

**Comment by raghavsethi** [comment]
Merged.
